src/pybind/mgr/cephadm/module.py

```diff
@@ -1920,65 +1935,91 @@ def _remove_daemon(self, name, host, force=False):
         self.cache.invalidate_host_daemons(host)
         return "Removed {} from host '{}'".format(name, host)
 
-    def _update_service(self, daemon_type, add_func, spec):
-        daemons = self.cache.get_daemons_by_type(daemon_type)
-        if len(daemons) > spec.count:
+    def _apply_service(self, daemon_type, spec, create_func, config_func=None):
+        """
+        Schedule a service.  Deploy new daemons or remove old ones, depending
+        on the target label and count specified in the placement.
+        """
+        service_name = daemon_type
+        if spec.name:
+            service_name += '.' + spec.name
+        daemons = self.cache.get_daemons_by_service(service_name)
+        spec = NodeAssignment(
```

Comment:
Would it make sense to teach `NodeAssignment` this logic? Strictly speaking we're assigning nodes for placement here.

Reply:
Probably?  Part of the goal here is to get to a single NodeAssignment caller/user so that we can expand/improve.  It needs to take into considering not just the node labels but also the placement of (IMO) *all* other daemons in the cluster.  I was considering that the next step once this PR goes in